bbbc022/first-hits-single-cell.py

```diff
@@ -0,0 +1,166 @@
+    filename = PROJECT_ROOT + "outputs/" + EXP + "/features/{}/{}/{}.npz"
+    filename = filename.format(
+        meta.loc[i, "Metadata_Plate"], 
+        meta.loc[i, "Metadata_Well"], 
+        meta.loc[i, "Metadata_Site"],
+    )
+    if os.path.isfile(filename):
+        with open(filename, "rb") as data:
+            info = np.load(data)
+            cells = info["features"].shape[0]
+            sc_meta_idx += [i]*cells
+            for c in range(cells):
+                sc_sc_idx.append(c)
+            total_single_cells += cells
+            
+
+cols = ["Metadata_Plate","Metadata_Well","Metadata_Site","Plate_Map_Name","broad_sample_Replicate","Treatment","Compound","Concentration","Replicate_Use"]
+sc_meta = pd.merge(pd.DataFrame(sc_meta_idx, columns=["ID"]), meta[cols], left_on="ID", right_index=True)
+sc_meta = pd.merge(pd.DataFrame(sc_sc_idx, columns=["SCID"]), sc_meta, left_index=True, right_index=True)
+
+sc_controls = sc_meta[sc_meta[PERT_NAME] == CTRL_NAME]
+sc_treatments = sc_meta[sc_meta[PERT_NAME] != CTRL_NAME]
+
+feats = [i for i in range(NUM_FEATURES)]
+
+def request_controls(control_meta_subset):
+    control_meta_subset.reset_index(inplace=True, drop=True)
+    sc_control_features_array = np.zeros((control_meta_subset.shape[0] , NUM_FEATURES))
+    for i in tqdm(control_meta_subset.index):
+        filename = PROJECT_ROOT + "outputs/" + EXP + "/features/{}/{}/{}.npz"
+        filename = filename.format(
+            control_meta_subset.loc[i, "Metadata_Plate"], 
+            control_meta_subset.loc[i, "Metadata_Well"], 
+            control_meta_subset.loc[i, "Metadata_Site"],
+        )
+        if os.path.isfile(filename):
+            with open(filename, "rb") as data:
+                info = np.load(data)
+                sc_control_features_array[i,:] = info["features"][control_meta_subset.loc[i, "SCID"]]
+    
+    return pd.merge(control_meta_subset, pd.DataFrame(data=sc_control_features_array), how="left", left_index=True, right_index=True)
+
+
+def request_treatments(treatments_meta_subset):
+    treatments_meta_subset.reset_index(inplace=True, drop=True)
+    sc_treatment_features_array = np.zeros((treatments_meta_subset.shape[0] , NUM_FEATURES))
+    for i in tqdm(treatments_meta_subset.index):
+        filename = PROJECT_ROOT + "outputs/" + EXP + "/features/{}/{}/{}.npz"
+        filename = filename.format(
+            treatments_meta_subset.loc[i, "Metadata_Plate"], 
+            treatments_meta_subset.loc[i, "Metadata_Well"], 
+            treatments_meta_subset.loc[i, "Metadata_Site"],
+        )
+        if os.path.isfile(filename):
+            with open(filename, "rb") as data:
+                info = np.load(data)
+                sc_treatment_features_array[i,:] = info["features"][treatments_meta_subset.loc[i, "SCID"]]
+    
+    return pd.merge(treatments_meta_subset, pd.DataFrame(data=sc_treatment_features_array), how="left", left_index=True, right_index=True)
+
+
+for i in range(10):
```

Comment:
Bring all the main code to this block and wrap it in the following conditional:
```
if __name__ == "__main__":
    # Executable code here
```
See why here: https://stackoverflow.com/questions/419163/what-does-if-name-main-do